src/utils.ts

```diff
@@ -0,0 +1,118 @@
+import { breakpoint } from '@core-ds/primitives';
+import { css } from 'styled-components';
+import { Breakpoint, EventKeys, ResponsiveStyle, Style } from './types';
+
+const breakpointKeys: Array<keyof Breakpoint> = Object.keys(breakpoint) as any;
+
+export function isResponsiveStyle<Value>(x?: Style<Value>): x is ResponsiveStyle<Value> {
+   return x != null && typeof x !== 'string';
+}
+
+export const responsiveProperty = (breakpointKey: string, name: string, propertyName: string) =>
+   css<any>`
+      ${(props) =>
+         isResponsiveStyle(props[name]) &&
+         props[name][breakpointKey] &&
+         css`
+            ${propertyName}: ${props[name][breakpointKey]};
+         `}
+   `;
+
+export const baseProperty = (propName: string, cssPropertyName: string) =>
+   css<any>`
+      ${(props) =>
+         isResponsiveStyle(props[propName]) && props[propName].base
+            ? css`
+                 ${cssPropertyName}: ${props[propName].base};
+              `
+            : props[propName] == null
+            ? null
+            : css`
+                 ${cssPropertyName}: ${props[propName]};
+              `}
+   `;
+
+export const createResponsiveStyles = (breakpointMapper: (key: keyof Breakpoint) => any) => {
+   return breakpointKeys.map(
+      (key) => css`
+         @media screen and (min-width: ${breakpoint[key]}) {
+            ${breakpointMapper(key)}
+         }
+      `
+   );
+};
+
+export const layout = css`
+   ${baseProperty('margin', 'margin')};
+   ${baseProperty('marginX', 'margin-left')};
+   ${baseProperty('marginX', 'margin-right')};
```

Comment:
I had this idea for how we could make these types of declarations a bit less repetitive. They're already a lot better than if we had to write it out the old fashioned way, but in the interest of reducing boilerplate I wanted to try condensing it a bit. I haven't tested this, but I think it should work. Not sure how the performance compares. I tried to make it easy to define the prop to css mapping once without having to make a separate call for base and responsive styles. I think with a little effort, we could infer the types based on this mapping, too, but I'm not really sure if it's possible. Let me know what you think.

```ts
// propName => cssProperty[]
const layoutCssMap = {
   margin: ['margin'],
   marginX: ['margin-left', 'margin-right'],
   ...
};

const gridCssMap = {
   templateRows: ['grid-template-rows'],
   ...
};

const makeCss = (cssMap) => (props) => {
   const allValues = Object.keys(cssMap).filter(propName => props[propName]).flatMap(propName => {
      const normalizedStyles = isResponsiveStyle(props[propName])
         ? props[propName]
         : {base: props[propName]};
      return cssMap[propName].flatMap(cssProperty => {
         const baseStyle = normalizedStyles.base ? `${cssProperty}: ${normalizedStyles.base};` : null;
         const responsiveStyles = breakpointKeys.filter(bpKey => normalizedStyles[bpKey]).map(bpKey => 
            `@media screen and (min-width: ${breakpoint[bpKey]}) {
               ${cssProperty}: ${normalizedStyles[bpKey]};
            }`
         );
         return [baseStyle, ...responsiveStyles];
      });
   });
   return allValues.join("\n");
};

const Grid = styled.grid<GridProps>`
   ${makeCss(gridCssMap)}
   ${makeCss(layoutCssMap)}
`;

Reply:
The API looks nice! I haven't tested this, but before we dive into this path I suggest we consider using [Styled System](https://styled-system.com/). It basically implements every css props and can help you enforce your core primitives. It's also completely compatible with styled components.

I haven't suggested it immediately because I think that it's worth using it only if it is widely adopted across the codebase since it doesn't really make sense to add the dependency if it's used only for the footer. For this project I threw there a quick and dirty implementation of this concept, but if you like this approach I strongly recommend we use Styled System instead.

Reply:
Yeah, I'm interested in Styled System! Some of the other libraries based on it like theme-ui look interesting too.